clang/test/CodeGenSYCL/stdtypes_kernel_type.cpp

```diff
@@ -0,0 +1,63 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -fsycl-int-header=%t.h -DCHECK_ERROR -verify %s
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -fsycl-int-header=%t.h %s
+// RUN: FileCheck -input-file=%t.h %s
+//
+// CHECK: #include <CL/sycl/detail/defines.hpp>
+// CHECK-NEXT: #include <CL/sycl/detail/kernel_desc.hpp>
+//
+// CHECK: static constexpr
+// CHECK-NEXT: const char* const kernel_names[] = {
+// CHECK-NEXT:   "_ZTSm",
+// CHECK-NEXT:   "_ZTSl"
+// CHECK-NEXT: };
+//
+// CHECK: static constexpr
+// CHECK-NEXT: const kernel_param_desc_t kernel_signatures[] = {
+// CHECK-NEXT:   //--- _ZTSm
+// CHECK-EMPTY:
+// CHECK-NEXT:   //--- _ZTSl
+// CHECK-EMPTY:
+// CHECK-NEXT: };
+//
+// CHECK: static constexpr
+// CHECK-NEXT: const unsigned kernel_signature_start[] = {
+// CHECK-NEXT:  0, // _ZTSm
+// CHECK-NEXT:  1 // _ZTSl
+// CHECK-NEXT: };
+
+// CHECK: template <> struct KernelInfo<unsigned long> {
+// CHECK: template <> struct KernelInfo<long> {
+
+void usage() {
+}
+
+namespace std {
+typedef long unsigned int size_t;
+typedef long int ptrdiff_t;
+typedef decltype(nullptr) nullptr_t;
+class T;
+class U;
+} // namespace std
+
+template <typename T>
+struct Templated_kernel_name;
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+#ifdef CHECK_ERROR
+  kernel_single_task<std::nullptr_t>([=]() {});                        // expected-error {{kernel name cannot be a type in the "std" namespace}}
+  kernel_single_task<std::T>([=]() {});                                // expected-error {{kernel name cannot be a type in the "std" namespace}}
+  kernel_single_task<Templated_kernel_name<std::nullptr_t>>([=]() {}); // expected-error {{kernel name cannot be a type in the "std" namespace}}
```

Comment:
I think this should be allowed. The KernelName here is `Templated_kernel_name<std::nullptr_t>` This is not in the 'std' namespace. I do not think there is any spec restriction which says we cannot template a kernel with types from std namespace. 

I think this is why the tests are failing as well. Changing this will probably fix the tests without the need for 'SpecializationKernelName'. I think this change might be breaking valid code

Reply:
The problem here is that the justification here is the same.  Whether you do kernel_single_task<std::string> or kernel_single_task<some_template<std::string>>, we cannot actually put it in the int-header, because we cannot forward declare std::string.

nullptr_t and std::size_t/ptrdiff_t are less troublesome, since we know they are well defined typedefs (and can just spell them that way), but other types we cannot.

Reply:
> the standard doesn't allow us to forward declare anything in `std`.

@jtmott-intel, if a Kernel Name is not a std-type itself, but is templated on a std-type instead, like: foo<std::some_type>, what is the case where we need to forward-declare the std-type?   We only need to forward-declare foo, right?

Reply:
@premanadrao: How would we get away with that?  We cannot spell foo<std::string> without spelling std::string.

Reply:
In current code, I think std::some_type is also forward declared. There is no check on template parameters IIRC. However, if std header is included prior to integration header, I see no reason why we need to forward declare std::sometype

Reply:
> @premanadrao: How would we get away with that? We cannot spell foo<std::string> without spelling std::string.

Ah, because the integration header does not necessarily follow the included std header.   Got it.

Reply:
> In current code, I think std::some_type is also forward declared. 
The C++ standard prohibits doing this.

As I think about it, we could potentially FORCE the int-header to include some of the std:: headers, but that would make compile times even more interesting... Particularly because we don't seem to have a good way to inject the includes into the int-headers after the fact.

Reply:
We could include exactly those std:: headers which the original source file includes. That shouldn't add anything to the compile time because they should all have proper include guards.

Off topic for this PR (which is focused on std types): This doesn't solve the wider problem. A user can have other library types which aren't fwd-declarable. But for any arbitrary library headers we can't include them from the int-header because we don't know whether they have proper include guards.

Reply:
There's a lot of interesting ideas here. Including the headers that the original source includes could remove any need for forward declarations and all the problems that go with it. So my next question is, should we abandon this PR? This PR tweaks the existing forward declaration approach.

Reply:
Including headers from the int-header will likely cause all kinds of problems. We would effectively change the include order the user wrote. This will mess with e.g. precompiled headers, system headers which use preprocessor switches (e.g. `__USE_FILE_OFFSET64` or `WIN32_LEAN_AND_MEAN`) and user/library headers which redefines macros.

Reply:
OK. If after this discussion, we're deciding to stay with the existing forward declaration approach, then is this PR okay to merge? I'll need fresh approvals from code owners. @elizabethandrews, @Fznamznon, @premanandrao

Reply:
I still think we have to prohibit everything in namespace 'std', with some allowance for the types where desugaring/canonicalizing results in a non-std type (basically, anything that was typedef'ed to a builtin type).

Reply:
@erichkeane I think that's what we have now in this PR. The [`if (NS->isStdNamespace()) {`](https://github.com/intel/llvm/pull/1963/files#diff-c88a590aca6d5f8e92dd1f8b6ec05a49R2273) prohibits everything in namespace std. Some types such as `std::size_t` desugar to a builtin type such as `unsigned long` and are still allowed to work. Only `nullptr_t` is singled out, since `std::nullptr_t` also desugars to a builtin type but one that we explicitly check for and disallow. Is there a specific case you had in mind that should be changed?

Reply:
No just being clear.  I wouldn't mind allowing std::nullptr_t as well, since we can just canonicalize it as decltype(nullptr_t) though.

Reply:
I don't know what to do here. Things need to be forward declarable for us to do the right thing in the int-header. std types aren't. So we don't really have a choice. But we have codes using std-types in the name. Therefore if we go ahead with this solution we'll break codes. And the workaround shown above (with `SpecializationKernelName_XYZ`) doesn't work for someone who needs to write a kernel which works with any type. They can't provide a custom name for each without asking their users to provide a name. I hope @mkinsner has an idea.

Reply:
Just talked with @mkinsner . The intention is that that names and their template arguments are forward declarable. This excludes std types. We think it is OK for us to not support names with std names as template arguments. For those use cases users should use unnamed lambdas.